test/types/union_schema.test-d.ts

```diff
@@ -0,0 +1,45 @@
+import { expectType, expectError, expectNotType, expectNotAssignable } from 'tsd';
+
+import { Collection } from '../../src/collection';
+import { ObjectId } from '../../src/bson';
+import { Db } from '../../src/db';
+import { MongoClient } from '../../src/mongo_client';
+
+type InsertRes<TId = ObjectId> = Promise<{ acknowledged: boolean; insertedId: TId }>;
+
+const db = new Db(new MongoClient(''), '');
+
+interface Circle {
+  _id: ObjectId;
+  radius: number;
+}
+interface Rectangle {
+  _id: ObjectId;
+  height: number;
+  width: number;
+}
+type Shape = Circle | Rectangle;
+
+type c = Collection<Shape>;
+// should not insert a portion of a type
+expectNotAssignable<Parameters<c['insertOne']>[0]>({ height: 2 });
```

Comment:
I cleaned up the tests quite a bit here. When I was first playing with tsd I was writing code that I would check the expectType against but you can cut out a lot of boilerplate if you extract say the type of the first argument to `insertOne` so I've tried to do that more here. Grabbing the return type of findOne however is a bit more difficult since it changes depending upon the arguments passed in, so it is easier to write out some fake findOne calls to check the results.

Turns out the mixed properties between unions is permitted, you can see the added test. I'm not sure if this is something to be disallowed or not 🤔

Reply:
We discussed adding a test for a true XOR setup as a sanity check to make sure our added type inference doesn't break it;

we also discussed adding some of our sample code (translating existing js to ts) as an explicit typescript compilation test to make sure users trying to use our samples don't immediately run into issues;

ideally, we also talked about potentially adding a TS compilation test for each public constructor/API to make sure they are returning expected types for each of a set of use cases - I recommend splitting this off into a separate ticket as an immediate follow up to this one

Reply:
Added the test for the XOR union type, it does maintain its restrictions.